### PR TITLE
Fix recursive gen attempt 2

### LINKED ISSUE #452: On fixing Malli's recursive generation
For validation, current support for recursive data via `:ref` seems to be appropriate and satisfactory. The problem is that this design is not fit for generation. Any example, even a trivial one, becomes so impractical that a size of 2 or 3 has to be used. Quite often, it is already too high (and one cannot go much lower than that). For comparison, default size is 30 and defaut maximum size during tests is 200. Yet, Malli is an excellent library and `test.check` does provide a pretty good solution to what I am about to describe. The hard part is to solve how they can fit together without disturbing (too much) the current API.

In my experience, solving this issue would really push Malli forwards by a lot.

This is what we can call a problem of "exponential explosion". The fact that the size of some recursive data is not controlled as it gets generated. It means that instead of generating recursive data that is somewhat linear to the given size, it is rather somewhat exponential. Let us take this example from the README:

```clojure
[:schema {:registry {::cons [:maybe [:tuple pos-int? [:ref ::cons]]]}}
  ::cons]
```
This example can be generated easily but only because size has no influence on it, the recursive case depends on `:maybe`. More often, the recursive case does depend on the size. Let us add a `:vector`:

```clojure
[:schema {:registry {::cons [:maybe [:vector [:tuple pos-int? [:ref ::cons]]]]}}
  ::cons]
```

Now, with a default size of 30, generated values are often absolutely massive. So you can easily imagine how this is troublesome with any real world example. This is because every time a vector is generated, by chance, any of its element may be another vector generated with the same size as its parent, and so on.

Some cases can be mitigated by #449, by biasing generation towards leaves (non-recursive items) rather than containers (recursive items, collections of leaves). It is an easy fix and it is useful for other use cases anyway. However, this is not enough. It can mitigate our problem, and sometimes recursive generation needs to be biased anyway, but it is not a general solution.

What is missing is a mechanism that generate leaves with the given size but controls and diminishes the given size for containers. Let us take that kind of definition, very common in my experience:

```clojure
[:schema
 {:registry {::data    [:or
                        ::int
                        ::vector]
             ::int     :int
             ::vector  [:vector
                        [:ref ::data]]}}
 ::data]
```

With a size of 30, I often need roughly 10 screens to display a typical generated structure. An actual use case would often have a few leaves besides `::int` and a few containers besides `::vector`. As we see, there is no distinction between the leaf and the container (unless you traverse `:or` children and notice `::vector` cycles back to it, but that is not great for performance). If we had a way for making such a distinction, then we could easily leverage `clojure.test.check.generators/recursive-gen`: https://github.com/clojure/test.check/blob/test.check-0.10.0/src/main/clojure/clojure/test/check/generators.cljc#L1641

Their solution is really quite good and anyway, I don't think there is any other way to really outsmart it without deeply modifying `test.check`. I have a bit of a hard time imagining how current `:ref` alone could leverage this. It looks like it cannot do much on its own, without context. For performance reasons, I guess a new kind of schema for specifying containers and leaves would be a lot easier to implement. On the other hand, the API might become confusing. Really, I have no firm opinion at the moment besides the fact that we should somehow have support for `recursive-gen` alongside #449.

### PR DESCRIPTION
**Pull request by frenchy64**
Close #452

Another attempt at https://github.com/metosin/malli/pull/507

### CODE REVIEW

**Comment by frenchy64** [comment]
@ikitommi here's another shot at the recursive gen fix, all changes are in the generators ns. Still need a few more tests, but what do you think?

**Comment by ikitommi** [comment]
I finally had time to read this through, looks really good as everything is resolved on the `generator` ns. Big thanks to your efforts on this, I really appreciate. Before merging, you think this is done, adds real support for recursive schemas with no new gotchas for _using_ it? e.g. `TODO test for :ref shadowing:`, I would not know how to handle the shadowing if that would not worked (and if that ever worked before).

File: src/malli/generator.cljc [hunk 1]
```diff
@@ -6,6 +6,7 @@
             [clojure.test.check.properties :as prop]
             [clojure.test.check.random :as random]
             [clojure.test.check.rose-tree :as rose]
+            [clojure.spec.gen.alpha :as ga]
```
Comment (ikitommi) [inline src/malli/generator.cljc:9]: double import of this.

File: src/malli/generator.cljc [hunk 2]
```diff
@@ -45,62 +88,91 @@
     (cond
       (and min (= min max)) (gen/fmap str/join (gen/vector gen/char min))
       (and min max) (gen/fmap str/join (gen/vector gen/char min max))
+      ;;FIXME (* 2 0) = 0
```
Comment (ikitommi) [inline src/malli/generator.cljc:91]: good catch!

**Comment by ikitommi** [comment]
oh, missed the `Still need a few more tests` part. I'll merge this when you are happy with it.

**Comment by frenchy64** [comment]
@ikitommi thanks for looking. Adding the test indeed revealed a bug, and I don't think my solution is general enough. Needs more thought.

**Comment by frenchy64** [comment]
I think I figured it out--very subtle! Explained the reasoning in comments.

I will try and improve the error message for infinite schemas, right now it's pretty confusing.

**Comment by frenchy64** [comment]
Still needs more work. I broke something: the generation never goes more than 1 recursion in the example in the original issue:

```clojure
(mg/sample
  [:schema {:registry {::cons [:maybe [:tuple pos-int? [:ref ::cons]]]}}
   ::cons]
  {:size 30})
=> (nil [2 nil] [2 nil] [2 nil] [4 nil] [7 nil] [3 nil] nil [1 nil] nil [1 nil] [17 nil] [490 nil] [3444 nil] nil nil nil [20236 nil] nil nil [501 nil] nil [75 nil] [2703 nil] nil [14 nil] [237664 nil] nil [3733313 nil] [2 nil])
```

**Comment by frenchy64** [comment]
The problem was because caching does not take into account `options`. We need to disable caching while generating recursive generators. Added a test to show that recursive generators still cache.

I will sleep on it.

**Comment by ikitommi** [comment]
Thanks for working on this.

File: src/malli/generator.cljc (hunk 2 above)
Comment (frenchy64) [reply src/malli/generator.cljc]: Fixed via:

```
      min (gen/fmap str/join (gen/sized (fn [size] (gen/vector gen/char min (clojure.core/max size min)))))
```

Added test `malli.generator-test/generator-min-only-test`.

**Comment by frenchy64** [comment]
@ikitommi thanks. If you're happy with it, please merge. Happy to address any feedback.

> Before merging, you think this is done, adds real support for recursive schemas with no new gotchas for using it?

There's one optional tweak that users can make to their schemas to optimize their generators---I've added that to the README. Completely optional though.

The only people required to make changes is anyone that has created their own schemas that use `-recur` or `-maybe-recur`. They should trigger the `println`'s I've added in those functions and hopefully they will be changed as a result.

**Comment by ikitommi** [comment]
This PR is superb in all ways I think of - it solves a hard problem in an elegant way with easy-to-follow explanation of what happens and why, lot of tests to verify it works correctly while keeping the code in the `malli.generator` ns. I learned a lot of implementation by following the hand-crafted generators in tests.

Big thanks @frenchy64!

**Pull request merged**